### PR TITLE
Adopt changes to Maven resolution logic from core

### PR DESCRIPTION
**Pull request by proksch**
This PR introduces a couple of changes that are required to adopt the changes to the Maven data structures and the new resolution framework from core:

- Move Maven data structures to core and adopt new resolution API
- Provide an incremental `IMavenResolver` that feeds through Kafka and caches results on disk
- Provide (still widely untested) proof-of-concept implementation for Maven resolution micro-service
- Add Kafka extension to control the auto-commit behavior via CLI argument(*)
- Fix a configuration mistake in Kafka that lead to auto-commit behavior
- Move Jackson ObjectMapper handling to core package
- Extend IoUtils to simplify writing .json and .zip files
- Extend clean-up logic that can detect download errors in .m2 folder

(*) From the Kafka side, auto-commit is always `false`, as we manually commit in `KafkaImpl`. The difference that this change makes is how the `KafkaImpl` class handles commits (whether it commits after ever message or requires a method call). This is mainly used in the `depgraph` implementation to ONLY update the offset AFTER writing the processed messages to disk.

### CODE REVIEW

File: plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/DependencyGraphResolution.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+public class DependencyGraphResolution {
+
+    private IMavenResolver resolver;
+
+    @Inject
+    public DependencyGraphResolution(IMavenResolver resolver) {
+        this.resolver = resolver;
+    }
+
+    // TODO return a "Response" that allows us to add error information
```
Comment (mir-am) [inline plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/DependencyGraphResolution.java:45]: Should this TODO be addressed in this PR?

File: plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/Main.java [hunk 2]
```diff
@@ -15,55 +15,163 @@
...
+import eu.fasten.core.maven.data.Pom;
+import eu.fasten.core.maven.resolution.MavenResolverData;
 
 public class Main implements Plugin {
 
     private static final Logger LOG = LoggerFactory.getLogger(Main.class);
+    private static final int NUM_TO_REPORT = 1000;
+
+    // TODO this thresholds still need tweaking
+    private static final int MIN_STORAGE_NUMBER = 10000;
```
Comment (mir-am) [inline plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/Main.java:47]: Maybe you can introduce a CLI arg for the `MIN_STORAGE_NUMBER` threshold with a default value.

File: plugins/pom-analyzer/src/main/java/eu/f4sten/pomanalyzer/PomAnalyzerInjectorConfig.java [hunk 3]
```diff
@@ -51,9 +48,4 @@ public DatabaseUtils bindDatabaseUtils(PostgresConnector pc, JsonUtils json, Ver
         var dslContext = DSL.using(c, SQLDialect.POSTGRES);
         return new DatabaseUtils(dslContext, json, version);
     }
-
-    @ProvidesIntoSet
-    public Module bindJacksonModule() {
```
Comment (mir-am) [inline plugins/pom-analyzer/src/main/java/eu/f4sten/pomanalyzer/PomAnalyzerInjectorConfig.java:56]: Might be a stupid question! Why is this removed?

File: plugins/pom-analyzer/src/main/java/eu/f4sten/pomanalyzer/utils/DatabaseUtils.java [hunk 4]
```diff
@@ -82,7 +82,7 @@ public void insertIntoDB(PomAnalysisResult r, MetadataDao dao) {
             var depProduct = dep.groupId + Constants.mvnCoordinateSeparator + dep.artifactId;
             final var depId = dao.insertPackage(depProduct, Constants.mvnForge);
             var json = jsonUtils.toJson(dep);
-            dao.insertDependency(packageVersionId, depId, dep.getVersionConstraints(), null, null, null, json);
+            dao.insertDependency(packageVersionId, depId, dep.getVersionConstraintsArr(), null, null, null, json);
```
Comment (mir-am) [inline plugins/pom-analyzer/src/main/java/eu/f4sten/pomanalyzer/utils/DatabaseUtils.java:85]: This method is deprecated. Maybe we should remove the deprecation annotation from the method in `core`.

File: plugins/pom-analyzer/src/test/java/eu/f4sten/pomanalyzer/MainTest.java [hunk 5]
```diff
@@ -65,9 +60,9 @@ public void setup() {
 
         sut = new Main(tracker, repo, modelBuilder, extractor, db, resolver, kafka, args, msgs, fixer);
 
-        when(extractor.process(eq(null))).thenReturn(new PomAnalysisResult());
-        when(extractor.process(any(Model.class))).thenReturn(new PomAnalysisResult());
-        when(fixer.checkPackage(any(PomAnalysisResult.class))).thenReturn("jar");
+//        when(extractor.process(eq(null))).thenReturn(new Pom());
```
Comment (mir-am) [inline plugins/pom-analyzer/src/test/java/eu/f4sten/pomanalyzer/MainTest.java:63]: Why are these lines commented out?

**Review by mir-am — APPROVED** [review]
Nice work!
I have some minor comments.

File: plugins/pom-analyzer/src/main/java/eu/f4sten/pomanalyzer/PomAnalyzerInjectorConfig.java (hunk 3 above)
Comment (proksch) [reply plugins/pom-analyzer/src/main/java/eu/f4sten/pomanalyzer/PomAnalyzerInjectorConfig.java:56]: The module was empty.

File: plugins/pom-analyzer/src/main/java/eu/f4sten/pomanalyzer/utils/DatabaseUtils.java (hunk 4 above)
Comment (proksch) [reply plugins/pom-analyzer/src/main/java/eu/f4sten/pomanalyzer/utils/DatabaseUtils.java:85]: Agree. This seems to be part of a larger refactoring though that we should do. Maybe we could leave it out from this PR.

File: plugins/pom-analyzer/src/test/java/eu/f4sten/pomanalyzer/MainTest.java (hunk 5 above)
Comment (proksch) [reply plugins/pom-analyzer/src/test/java/eu/f4sten/pomanalyzer/MainTest.java:63]: The test was not running any meaningful test, but only served me as a basis for running debugging sessions. The changes broke the interfaces and it was easier to just comment these lines. I will fix it next time I need to debug the `PomAnalyzer` or when we finally decide to write tests for it. ^^

File: plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/DependencyGraphResolution.java (hunk 1 above)
Comment (proksch) [reply plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/DependencyGraphResolution.java:45]: No, this is just a note to future me.

File: infrastructure/infrastructure-impl/src/main/java/eu/f4sten/infra/impl/kafka/KafkaConnector.java [hunk 6]
```diff
@@ -84,6 +86,8 @@ public Properties getConsumerProperties(Lane l) {
         p.setProperty(VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
         p.setProperty(FETCH_MAX_BYTES_CONFIG, MAX_REQUEST_SIZE);
         p.setProperty(MAX_POLL_RECORDS_CONFIG, "1");
+        p.setProperty(ENABLE_AUTO_COMMIT_CONFIG, "false");
```
[the remark this answers is not in the document]
Comment (proksch) [reply infrastructure/infrastructure-impl/src/main/java/eu/f4sten/infra/impl/kafka/KafkaConnector.java:89]: No, these are two different settings. From the Kafka side, `auto-commit` should always be false, as we like to manually commit. The difference is how the `KafkaImpl` class handles it (whether it commits after ever message or not).

File: plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/Main.java (hunk 2 above)
Comment (proksch) [reply plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/Main.java:47]: I think this is a good idea, I will do that.

Comment (proksch) [reply plugins/dependency-graph-resolver/src/main/java/eu/f4sten/depgraph/Main.java]: Fixed in the latest commit. I have also updated the numbers based on my own experience after playing around with this on monster.

File: infrastructure/infrastructure-impl/src/main/java/eu/f4sten/infra/impl/kafka/KafkaConnector.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (proksch) [reply infrastructure/infrastructure-impl/src/main/java/eu/f4sten/infra/impl/kafka/KafkaConnector.java:89]: I have clarified it in the description of the PR as well to document it for others that stumble over this change.

**Review by mir-am — APPROVED** [review]
Thanks for addressing the comments and answering the questions in a short time.
No more comments from my side.
Please feel free to rebase and merge the PR.

**Pull request merged**